File: frontend/src/lib/types/product/product.ts
import {InputValueTypes} from "@/components/ui/input/types";
import {z} from "zod";

export interface Product {
    id: string;
    name: string;
    barcode: string;
    brand: string;
    price: number;
    category: string;
    lastUpdate: Date;
    stock: {
        currentQuantity: number;
        minThreshold: number;
        maxThreshold: number;
    };
    nutritionalInfo: {
        energy: number;
        proteins: number;
        carbohydrates: number;
        fat: number;
        fiber: number;
        ingredients: string;
    };
}

type NutrientLevels = null | 'low' | 'moderate' | 'high';
interface Nutrient {
    fat: NutrientLevels;
    saturatedFat: NutrientLevels;
    sugars: NutrientLevels;
    salt: NutrientLevels;
}

interface Nutriments {
    energyKcal100g: number;
    proteins100g: number;
    carbohydrates100g: number;
    fat100g: number;
    fiber100g: number;
    salt100g: number;
    sugars100g: number;
}

interface ImageSizes {
    en: string;
    fr: string;
}

interface SelectedImages {
    display: ImageSizes;
    small: ImageSizes;
    thumb: ImageSizes;
}

export interface ProductOFF {
    id: string | null;
    barcode: string;
    category: string | null;
    brand: string;
    name: string;
    ingredients: string;
    price: InputValueTypes | null;
    nutrientLevels: Nutrient;
    nutriments: Nutriments;
    nutriscoreGrade: 'a' | 'b' | 'c' | 'd' | 'e';
    selectedImages: SelectedImages;
    lastUpdate: string | null;
    stock: {
        currentQuantity: InputValueTypes;
        minThreshold: InputValueTypes;
        maxThreshold: InputValueTypes;
    };
}

export interface ProductResponse {
    id: string;
    barcode: string;
    category: string | null;
    brand: string;
    name: string;
    ingredients: string;
    price: InputValueTypes;
    nutrientLevels: Nutrient;
    nutriments: Nutriments;
    nutriscoreGrade: 'a' | 'b' | 'c' | 'd' | 'e';
    selectedImages: SelectedImages;
    lastUpdate: string | null;
    stock: {
        quantity: InputValueTypes;
        minThreshold: InputValueTypes;
        maxThreshold: InputValueTypes;
    };
}

export const updateProductSchema = z.object({
    name: z.string(),
    price: z.number(),
    quantity: z.number()
});

export type UpdateProductSchemaType = z.infer<typeof updateProductSchema>;

File: backend/src/main/java/com/trinity/product/model/Product.java
package com.trinity.product.model;

import java.time.LocalDateTime;
import java.util.UUID;

import org.hibernate.annotations.CreationTimestamp;
import org.hibernate.annotations.UpdateTimestamp;
import org.springframework.data.annotation.LastModifiedDate;

import com.trinity.common.domain.exception.BusinessRuleViolation;

import java.math.BigDecimal;

import jakarta.persistence.*;
import jakarta.validation.constraints.NotNull;
import jakarta.validation.constraints.Positive;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;

@Entity
@Table(name = "products")
@Data
@Builder
@AllArgsConstructor
@NoArgsConstructor
public class Product {

    @Id
    @GeneratedValue(strategy = GenerationType.UUID)
    private UUID id;

    @Column(nullable = false, unique = true, length = 13)
    private String barcode;

    @Column(length = 100)
    private String category;

    @Column(length = 100)
    private String brand;

    @Column(nullable = false)
    private String name;

    private String ingredients;

    @NotNull
    @Positive
    @Column(nullable = false, precision = 10, scale = 2)
    private BigDecimal price;

    @Embedded
    private NutrientLevels nutrientLevels;

    @CreationTimestamp
    private LocalDateTime createdAt;

    @UpdateTimestamp
    private LocalDateTime updatedAt;

    @Embeddable 
    @Data
    @Builder
    @AllArgsConstructor
    @NoArgsConstructor
    public static class NutrientLevels {
        private String fat;
        private String saturatedFat;
        private String sugars;
        private String salt;
    }

    @Embedded
    private Nutriments nutriments;

    @Embeddable
    @Data
    @Builder
    @AllArgsConstructor
    @NoArgsConstructor
    public static class Nutriments {
        private double energyKcal100g;
        private double proteins100g;
        private double carbohydrates100g;
        private double fat100g;
        private double fiber100g;
        private double salt100g;
        private double sugars100g;
    }

    private String nutriscoreGrade; 

    @Embedded
    private SelectedImages selectedImages;

    @Embeddable
    @Data
    @Builder
    @AllArgsConstructor
    @NoArgsConstructor
    public static class SelectedImages {

        @Embedded
        @AttributeOverrides({
            @AttributeOverride(name = "en", column = @Column(name = "display_en")),
            @AttributeOverride(name = "fr", column = @Column(name = "display_fr"))
        })
        private ProductImageUrl display;

        @Embedded
        @AttributeOverrides({
            @AttributeOverride(name = "en", column = @Column(name = "small_en")),
            @AttributeOverride(name = "fr", column = @Column(name = "small_fr"))
        })
        private ProductImageUrl small;

        @Embedded
        @AttributeOverrides({
            @AttributeOverride(name = "en", column = @Column(name = "thumb_en")),
            @AttributeOverride(name = "fr", column = @Column(name = "thumb_fr"))
        })
        private ProductImageUrl thumb;
    }

    @LastModifiedDate
    private LocalDateTime lastUpdate;

   @Embedded
    private Stock stock;

    @Embeddable
    @Data
    @Builder
    @AllArgsConstructor
    @NoArgsConstructor
    public static class Stock {
        private int quantity;
        private int minThreshold;
        private int maxThreshold;

        public void reduceBy(int amount) {
            if (amount < 0) {
                throw new BusinessRuleViolation("Reduction amount must not be negative");
            }
            if (amount > this.quantity) {
                throw new BusinessRuleViolation("Stock quantity cannot become negative");
            }
            this.quantity -= amount;
        }

        public void increaseBy(int amount) {
            if (amount < 0) {
                throw new BusinessRuleViolation("Increase amount must not be negative");
            }
            this.quantity += amount;
        }

        public boolean isBelowThreshold() {
            return this.quantity < this.minThreshold;
        }
    }

    /* Domain behavior */

    public void changePrice(BigDecimal newPrice) {
        if (newPrice == null) {
            throw new BusinessRuleViolation("Price must not be null");
        }
        if (newPrice.compareTo(BigDecimal.ZERO) < 0) {
            throw new BusinessRuleViolation("Price cannot be negative");
        }
        this.price = newPrice;
    }

    /** Applies a signed delta to the stock, rejecting a resulting negative quantity. */
    public void adjustStock(int delta) {
        if (this.stock == null) {
            this.stock = Stock.builder().build();
        }
        if (delta >= 0) {
            this.stock.increaseBy(delta);
        } else {
            this.stock.reduceBy(-delta);
        }
    }
}